pennylane_sf/gbs.py

```diff
@@ -0,0 +1,258 @@
+
+    # pylint: disable=missing-function-docstring
+    def execute(self, queue, observables, parameters=None, **kwargs):
+        parameters = parameters or {}
+        if len(queue) > 1:
+            raise ValueError(
+                "The StrawberryFieldsGBS device accepts a single application of ParamGraphEmbed"
+            )
+        return super().execute(queue, observables, parameters, **kwargs)
+
+    # pylint: disable=pointless-statement,expression-not-assigned
+    def apply(self, operation, wires, par):
+        self._params, A, _ = par
+
+        if len(self._params) != self.num_wires:
+            raise ValueError(
+                "The number of variable parameters must be equal to the total number of wires."
+            )
+
+        self._WAW = self.calculate_WAW(*par)
+        n_mean_WAW = self.calculate_n_mean(self._WAW)
+
+        op = GraphEmbed(self._WAW, mean_photon_per_mode=n_mean_WAW / len(A))
+        op | [self.q[wires.index(i)] for i in wires]
+
+        if not self.analytic:
+            MeasureFock() | [self.q[wires.index(i)] for i in wires]
+
+    def pre_measure(self):
+        self.eng = sf.Engine("gaussian", backend_options={"cutoff_dim": self.cutoff})
+
+        if self.analytic:
+            results = self.eng.run(self.prog)
+        else:
+            results = self.eng.run(self.prog, shots=self.shots)
+
+        self.state = results.state
+        self.samples = results.samples
+
+    def probability(self, wires=None):
+        wires = wires or self.wires
+        wires = Wires(wires)
+
+        if self.analytic:
+            return super().probability(wires=wires)
+
+        N = len(wires)
+        samples = np.take(self.samples, self.wires.indices(wires), axis=1)
+
+        probs = all_fock_probs_pnr(samples)
+        ind = np.indices([self.cutoff] * N).reshape(N, -1).T
+
+        probs = OrderedDict((tuple(i), probs[tuple(i)]) for i in ind)
+        return probs
+
+    def jacobian(self, operations, observables, variable_deps):
+        """Calculates the Jacobian of the device.
+
+        Args:
+            operations (list[~pennylane.operation.Operation]): operations to be applied to the
+                device
+            observables (list[~pennylane.operation.Operation]): observables to be measured
+            variable_deps (dict[int, ParameterDependency]): reference dictionary
+                mapping free parameter values to the operations that
+                depend on them
+
+        Returns:
+            array[float]: Jacobian matrix of size (``len(probs)``, ``num_wires``)
+        """
+        self.reset()
+
+        requested_wires = observables[0].wires
+
+        # Create dummy observable to measure probabilities over all wires
+        obs_all_wires = qml.Identity(wires=self.wires)
+        obs_all_wires.return_type = Probability
+        prob = self.execute(operations, [obs_all_wires], parameters=variable_deps)[0]
+
+        jac = self._jacobian_all_wires(prob)
```

Comment:
Question: is it possible to use the marginal probabilities to compute the jacobian over requested wires? Or is it more efficient to do it this way? Just wondering :)